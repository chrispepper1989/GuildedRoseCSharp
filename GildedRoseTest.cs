using System;
using System.IO;
using System.Text;
using ApprovalTests;
using ApprovalTests.Reporters;
using csharp;

namespace GildedRoseCSharp
{
  
    [UseReporter(typeof(DiffReporter))]
    [TestFixture]
    public class GildedRoseTest
    {

        //generate a bunch of test cases
        public IList<Item> GenerateTestItems(ValidItems[] names, int[] sellIns, int[] qualities)
        {
            IList<Item> Items = new List<Item>
            {
            };
            foreach (var name in names)
            {
                foreach (var sellIn in sellIns)
                {
                    foreach (var quality in qualities)
                    {
                        var item = ItemHelper.CreateItem(name, sellIn, quality);
                        Items.Add( item);
                    }

                }

            }

            return Items;

        }

        //[TestCase(10)] //todo why doesnt this work with Approvals?
        public void TestUpdateQuality(int days)
        {

            //arrange
            ValidItems[] namesProper = new[]
            {
                ValidItems.AgedBrie, ValidItems.DexterityVest, ValidItems.ElixirOfMongoose,
                ValidItems.SulfurasHandOfRagnaros, ValidItems.ConjuredManaCake
            };
            var sellIns = new[] { 1, 5, 10, 2, 4 };
            var qualities = new[] { 0, 1, 2, 5, 29 };
            var Items = GenerateTestItems(namesProper, sellIns, qualities);
            GildedRose app = new GildedRose(Items);
            //act
            for (var i = 0; i < days; ++i)
            {
                app.UpdateQuality();
            }
            //assert
            Approvals.VerifyAll(Items, "All Items");

        }
        [Test]
        public void ConjuredManaCakeDegradesTwiceAsFastAsNormalItems()
        {
            //arrange
            var conitem = ItemHelper.CreateItem(ValidItems.ConjuredManaCake, 10, 100);
            var normalItem = ItemHelper.CreateItem(ValidItems.ElixirOfMongoose, 10, 100);

            IList<Item> Items = new List<Item>
            {
         
            };

            GildedRose app = new GildedRose(Items);

            IList<Item> ItemsOverTime = new List<Item>();
            //act
            for (var i = 0; i < 100; ++i)
            {
                conitem.SellIn -= 1;
                normalItem.SellIn -= 1;
                //normal item done twice
                normalItem.Quality =  GildedRose.GetNewStandardItemQuality(normalItem);
                normalItem.Quality = GildedRose.GetNewStandardItemQuality(normalItem);
                conitem.Quality = GildedRose.GetNewConjuredManaCakeQuality(conitem);
               
                //assert
                Assert.AreEqual(normalItem.Quality, conitem.Quality);
      
                ItemsOverTime.Add(new Item()
                {
                    Name = normalItem.Name,
                    Quality = normalItem.Quality,
                    SellIn = normalItem.SellIn

                });
                ItemsOverTime.Add(new Item()
                {
                    Name = conitem.Name,
                    Quality = conitem.Quality,
                    SellIn = conitem.SellIn

                });

            }
            Approvals.VerifyAll(ItemsOverTime, "Conjured Items ");

        }

        [Test]
         public void TestUpdateQuality_1Days()
         {
             TestUpdateQuality(1);
         }
         [Test]
         public void TestUpdateQuality_5Days()
         {
             TestUpdateQuality(5);
         }
         [Test]
         public void TestUpdateQuality_10Days()
         {
             TestUpdateQuality(10);
         }
         [Test]
         public void TestUpdateQuality_30Days()
         {
             TestUpdateQuality(30);
         }

    }
}
