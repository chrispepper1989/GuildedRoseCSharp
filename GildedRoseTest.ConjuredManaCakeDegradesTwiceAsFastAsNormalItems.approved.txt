Conjured Items [0] = Elixir of the Mongoose, 9, 98
Conjured Items [1] = Conjured Mana Cake, 9, 98
Conjured Items [2] = Elixir of the Mongoose, 8, 96
Conjured Items [3] = Conjured Mana Cake, 8, 96
Conjured Items [4] = Elixir of the Mongoose, 7, 94
Conjured Items [5] = Conjured Mana Cake, 7, 94
Conjured Items [6] = Elixir of the Mongoose, 6, 92
Conjured Items [7] = Conjured Mana Cake, 6, 92
Conjured Items [8] = Elixir of the Mongoose, 5, 90
Conjured Items [9] = Conjured Mana Cake, 5, 90
Conjured Items [10] = Elixir of the Mongoose, 4, 88
Conjured Items [11] = Conjured Mana Cake, 4, 88
Conjured Items [12] = Elixir of the Mongoose, 3, 86
Conjured Items [13] = Conjured Mana Cake, 3, 86
Conjured Items [14] = Elixir of the Mongoose, 2, 84
Conjured Items [15] = Conjured Mana Cake, 2, 84
Conjured Items [16] = Elixir of the Mongoose, 1, 82
Conjured Items [17] = Conjured Mana Cake, 1, 82
Conjured Items [18] = Elixir of the Mongoose, 0, 78
Conjured Items [19] = Conjured Mana Cake, 0, 78
Conjured Items [20] = Elixir of the Mongoose, -1, 74
Conjured Items [21] = Conjured Mana Cake, -1, 74
Conjured Items [22] = Elixir of the Mongoose, -2, 70
Conjured Items [23] = Conjured Mana Cake, -2, 70
Conjured Items [24] = Elixir of the Mongoose, -3, 66
Conjured Items [25] = Conjured Mana Cake, -3, 66
Conjured Items [26] = Elixir of the Mongoose, -4, 62
Conjured Items [27] = Conjured Mana Cake, -4, 62
Conjured Items [28] = Elixir of the Mongoose, -5, 58
Conjured Items [29] = Conjured Mana Cake, -5, 58
Conjured Items [30] = Elixir of the Mongoose, -6, 54
Conjured Items [31] = Conjured Mana Cake, -6, 54
Conjured Items [32] = Elixir of the Mongoose, -7, 50
Conjured Items [33] = Conjured Mana Cake, -7, 50
Conjured Items [34] = Elixir of the Mongoose, -8, 46
Conjured Items [35] = Conjured Mana Cake, -8, 46
Conjured Items [36] = Elixir of the Mongoose, -9, 42
Conjured Items [37] = Conjured Mana Cake, -9, 42
Conjured Items [38] = Elixir of the Mongoose, -10, 38
Conjured Items [39] = Conjured Mana Cake, -10, 38
Conjured Items [40] = Elixir of the Mongoose, -11, 34
Conjured Items [41] = Conjured Mana Cake, -11, 34
Conjured Items [42] = Elixir of the Mongoose, -12, 30
Conjured Items [43] = Conjured Mana Cake, -12, 30
Conjured Items [44] = Elixir of the Mongoose, -13, 26
Conjured Items [45] = Conjured Mana Cake, -13, 26
Conjured Items [46] = Elixir of the Mongoose, -14, 22
Conjured Items [47] = Conjured Mana Cake, -14, 22
Conjured Items [48] = Elixir of the Mongoose, -15, 18
Conjured Items [49] = Conjured Mana Cake, -15, 18
Conjured Items [50] = Elixir of the Mongoose, -16, 14
Conjured Items [51] = Conjured Mana Cake, -16, 14
Conjured Items [52] = Elixir of the Mongoose, -17, 10
Conjured Items [53] = Conjured Mana Cake, -17, 10
Conjured Items [54] = Elixir of the Mongoose, -18, 6
Conjured Items [55] = Conjured Mana Cake, -18, 6
Conjured Items [56] = Elixir of the Mongoose, -19, 2
Conjured Items [57] = Conjured Mana Cake, -19, 2
Conjured Items [58] = Elixir of the Mongoose, -20, 0
Conjured Items [59] = Conjured Mana Cake, -20, 0
Conjured Items [60] = Elixir of the Mongoose, -21, 0
Conjured Items [61] = Conjured Mana Cake, -21, 0
Conjured Items [62] = Elixir of the Mongoose, -22, 0
Conjured Items [63] = Conjured Mana Cake, -22, 0
Conjured Items [64] = Elixir of the Mongoose, -23, 0
Conjured Items [65] = Conjured Mana Cake, -23, 0
Conjured Items [66] = Elixir of the Mongoose, -24, 0
Conjured Items [67] = Conjured Mana Cake, -24, 0
Conjured Items [68] = Elixir of the Mongoose, -25, 0
Conjured Items [69] = Conjured Mana Cake, -25, 0
Conjured Items [70] = Elixir of the Mongoose, -26, 0
Conjured Items [71] = Conjured Mana Cake, -26, 0
Conjured Items [72] = Elixir of the Mongoose, -27, 0
Conjured Items [73] = Conjured Mana Cake, -27, 0
Conjured Items [74] = Elixir of the Mongoose, -28, 0
Conjured Items [75] = Conjured Mana Cake, -28, 0
Conjured Items [76] = Elixir of the Mongoose, -29, 0
Conjured Items [77] = Conjured Mana Cake, -29, 0
Conjured Items [78] = Elixir of the Mongoose, -30, 0
Conjured Items [79] = Conjured Mana Cake, -30, 0
Conjured Items [80] = Elixir of the Mongoose, -31, 0
Conjured Items [81] = Conjured Mana Cake, -31, 0
Conjured Items [82] = Elixir of the Mongoose, -32, 0
Conjured Items [83] = Conjured Mana Cake, -32, 0
Conjured Items [84] = Elixir of the Mongoose, -33, 0
Conjured Items [85] = Conjured Mana Cake, -33, 0
Conjured Items [86] = Elixir of the Mongoose, -34, 0
Conjured Items [87] = Conjured Mana Cake, -34, 0
Conjured Items [88] = Elixir of the Mongoose, -35, 0
Conjured Items [89] = Conjured Mana Cake, -35, 0
Conjured Items [90] = Elixir of the Mongoose, -36, 0
Conjured Items [91] = Conjured Mana Cake, -36, 0
Conjured Items [92] = Elixir of the Mongoose, -37, 0
Conjured Items [93] = Conjured Mana Cake, -37, 0
Conjured Items [94] = Elixir of the Mongoose, -38, 0
Conjured Items [95] = Conjured Mana Cake, -38, 0
Conjured Items [96] = Elixir of the Mongoose, -39, 0
Conjured Items [97] = Conjured Mana Cake, -39, 0
Conjured Items [98] = Elixir of the Mongoose, -40, 0
Conjured Items [99] = Conjured Mana Cake, -40, 0
Conjured Items [100] = Elixir of the Mongoose, -41, 0
Conjured Items [101] = Conjured Mana Cake, -41, 0
Conjured Items [102] = Elixir of the Mongoose, -42, 0
Conjured Items [103] = Conjured Mana Cake, -42, 0
Conjured Items [104] = Elixir of the Mongoose, -43, 0
Conjured Items [105] = Conjured Mana Cake, -43, 0
Conjured Items [106] = Elixir of the Mongoose, -44, 0
Conjured Items [107] = Conjured Mana Cake, -44, 0
Conjured Items [108] = Elixir of the Mongoose, -45, 0
Conjured Items [109] = Conjured Mana Cake, -45, 0
Conjured Items [110] = Elixir of the Mongoose, -46, 0
Conjured Items [111] = Conjured Mana Cake, -46, 0
Conjured Items [112] = Elixir of the Mongoose, -47, 0
Conjured Items [113] = Conjured Mana Cake, -47, 0
Conjured Items [114] = Elixir of the Mongoose, -48, 0
Conjured Items [115] = Conjured Mana Cake, -48, 0
Conjured Items [116] = Elixir of the Mongoose, -49, 0
Conjured Items [117] = Conjured Mana Cake, -49, 0
Conjured Items [118] = Elixir of the Mongoose, -50, 0
Conjured Items [119] = Conjured Mana Cake, -50, 0
Conjured Items [120] = Elixir of the Mongoose, -51, 0
Conjured Items [121] = Conjured Mana Cake, -51, 0
Conjured Items [122] = Elixir of the Mongoose, -52, 0
Conjured Items [123] = Conjured Mana Cake, -52, 0
Conjured Items [124] = Elixir of the Mongoose, -53, 0
Conjured Items [125] = Conjured Mana Cake, -53, 0
Conjured Items [126] = Elixir of the Mongoose, -54, 0
Conjured Items [127] = Conjured Mana Cake, -54, 0
Conjured Items [128] = Elixir of the Mongoose, -55, 0
Conjured Items [129] = Conjured Mana Cake, -55, 0
Conjured Items [130] = Elixir of the Mongoose, -56, 0
Conjured Items [131] = Conjured Mana Cake, -56, 0
Conjured Items [132] = Elixir of the Mongoose, -57, 0
Conjured Items [133] = Conjured Mana Cake, -57, 0
Conjured Items [134] = Elixir of the Mongoose, -58, 0
Conjured Items [135] = Conjured Mana Cake, -58, 0
Conjured Items [136] = Elixir of the Mongoose, -59, 0
Conjured Items [137] = Conjured Mana Cake, -59, 0
Conjured Items [138] = Elixir of the Mongoose, -60, 0
Conjured Items [139] = Conjured Mana Cake, -60, 0
Conjured Items [140] = Elixir of the Mongoose, -61, 0
Conjured Items [141] = Conjured Mana Cake, -61, 0
Conjured Items [142] = Elixir of the Mongoose, -62, 0
Conjured Items [143] = Conjured Mana Cake, -62, 0
Conjured Items [144] = Elixir of the Mongoose, -63, 0
Conjured Items [145] = Conjured Mana Cake, -63, 0
Conjured Items [146] = Elixir of the Mongoose, -64, 0
Conjured Items [147] = Conjured Mana Cake, -64, 0
Conjured Items [148] = Elixir of the Mongoose, -65, 0
Conjured Items [149] = Conjured Mana Cake, -65, 0
Conjured Items [150] = Elixir of the Mongoose, -66, 0
Conjured Items [151] = Conjured Mana Cake, -66, 0
Conjured Items [152] = Elixir of the Mongoose, -67, 0
Conjured Items [153] = Conjured Mana Cake, -67, 0
Conjured Items [154] = Elixir of the Mongoose, -68, 0
Conjured Items [155] = Conjured Mana Cake, -68, 0
Conjured Items [156] = Elixir of the Mongoose, -69, 0
Conjured Items [157] = Conjured Mana Cake, -69, 0
Conjured Items [158] = Elixir of the Mongoose, -70, 0
Conjured Items [159] = Conjured Mana Cake, -70, 0
Conjured Items [160] = Elixir of the Mongoose, -71, 0
Conjured Items [161] = Conjured Mana Cake, -71, 0
Conjured Items [162] = Elixir of the Mongoose, -72, 0
Conjured Items [163] = Conjured Mana Cake, -72, 0
Conjured Items [164] = Elixir of the Mongoose, -73, 0
Conjured Items [165] = Conjured Mana Cake, -73, 0
Conjured Items [166] = Elixir of the Mongoose, -74, 0
Conjured Items [167] = Conjured Mana Cake, -74, 0
Conjured Items [168] = Elixir of the Mongoose, -75, 0
Conjured Items [169] = Conjured Mana Cake, -75, 0
Conjured Items [170] = Elixir of the Mongoose, -76, 0
Conjured Items [171] = Conjured Mana Cake, -76, 0
Conjured Items [172] = Elixir of the Mongoose, -77, 0
Conjured Items [173] = Conjured Mana Cake, -77, 0
Conjured Items [174] = Elixir of the Mongoose, -78, 0
Conjured Items [175] = Conjured Mana Cake, -78, 0
Conjured Items [176] = Elixir of the Mongoose, -79, 0
Conjured Items [177] = Conjured Mana Cake, -79, 0
Conjured Items [178] = Elixir of the Mongoose, -80, 0
Conjured Items [179] = Conjured Mana Cake, -80, 0
Conjured Items [180] = Elixir of the Mongoose, -81, 0
Conjured Items [181] = Conjured Mana Cake, -81, 0
Conjured Items [182] = Elixir of the Mongoose, -82, 0
Conjured Items [183] = Conjured Mana Cake, -82, 0
Conjured Items [184] = Elixir of the Mongoose, -83, 0
Conjured Items [185] = Conjured Mana Cake, -83, 0
Conjured Items [186] = Elixir of the Mongoose, -84, 0
Conjured Items [187] = Conjured Mana Cake, -84, 0
Conjured Items [188] = Elixir of the Mongoose, -85, 0
Conjured Items [189] = Conjured Mana Cake, -85, 0
Conjured Items [190] = Elixir of the Mongoose, -86, 0
Conjured Items [191] = Conjured Mana Cake, -86, 0
Conjured Items [192] = Elixir of the Mongoose, -87, 0
Conjured Items [193] = Conjured Mana Cake, -87, 0
Conjured Items [194] = Elixir of the Mongoose, -88, 0
Conjured Items [195] = Conjured Mana Cake, -88, 0
Conjured Items [196] = Elixir of the Mongoose, -89, 0
Conjured Items [197] = Conjured Mana Cake, -89, 0
Conjured Items [198] = Elixir of the Mongoose, -90, 0
Conjured Items [199] = Conjured Mana Cake, -90, 0
